controllers/complianceeventsapi/types.go

```diff
@@ -0,0 +1,266 @@
+	for i, c := range p.Categories {
+		cats[i] = fmt.Sprintf("%q", c) // double-quoted and safely escaped with Go syntax
+	}
+
+	ctrls := make([]string, len(p.Controls))
+	for i, c := range p.Controls {
+		ctrls[i] = fmt.Sprintf("%q", c) // double-quoted and safely escaped with Go syntax
+	}
+
+	stands := make([]string, len(p.Standards))
+	for i, c := range p.Standards {
+		stands[i] = fmt.Sprintf("%q", c) // double-quoted and safely escaped with Go syntax
+	}
+
+	return parentPolicyKey{
+		Name:       p.Name,
+		Categories: fmt.Sprintf("[%v]", strings.Join(cats, ",")),
+		Controls:   fmt.Sprintf("[%v]", strings.Join(ctrls, ",")),
+		Standards:  fmt.Sprintf("[%v]", strings.Join(stands, ",")),
+	}
+}
+
+type parentPolicyKey struct {
+	Name       string
+	Categories string
+	Controls   string
+	Standards  string
+}
+
+type Policy struct {
+	KeyID          int     `db:"id" json:"-" goqu:"skipinsert"`
+	Kind           string  `db:"kind" json:"kind"`
+	APIGroup       string  `db:"api_group" json:"apiGroup"`
+	Name           string  `db:"name" json:"name"`
+	Namespace      *string `db:"namespace" json:"namespace,omitempty"`
+	ParentPolicyID *int    `db:"parent_policy_id" json:"-"`
+	Spec           *string `db:"spec" json:"spec,omitempty"`
+	SpecHash       *string `db:"spec_hash" json:"specHash,omitempty"`
+	Severity       *string `db:"severity" json:"severity,omitempty"`
+}
+
+func (p *Policy) Validate() error {
+	errs := make([]error, 0)
+
+	if p.APIGroup == "" {
+		errs = append(errs, fmt.Errorf("%w: policy.apiGroup", errRequiredFieldNotProvided))
+	}
+
+	if p.Kind == "" {
+		errs = append(errs, fmt.Errorf("%w: policy.kind", errRequiredFieldNotProvided))
+	}
+
+	if p.Name == "" {
+		errs = append(errs, fmt.Errorf("%w: policy.name", errRequiredFieldNotProvided))
+	}
+
+	if p.Spec == nil && p.SpecHash == nil {
+		errs = append(errs, fmt.Errorf("%w: policy.spec or policy.specHash", errRequiredFieldNotProvided))
+	}
+
+	if p.Spec != nil {
+		var buf bytes.Buffer
+		if err := json.Compact(&buf, []byte(*p.Spec)); err != nil {
+			errs = append(errs, fmt.Errorf("%w: policy.spec is not valid JSON: %w", errInvalidInput, err))
+		} else if buf.String() != *p.Spec {
+			errs = append(errs, fmt.Errorf("%w: policy.spec is not compact JSON", errInvalidInput))
+		} else if p.SpecHash != nil {
+			sum := sha1.Sum(buf.Bytes())
+
+			if *p.SpecHash != hex.EncodeToString(sum[:]) {
+				errs = append(errs, fmt.Errorf("%w: policy.specHash does not match the compact policy.Spec",
+					errInvalidInput))
+			}
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+func (p *Policy) key() policyKey {
+	key := policyKey{
+		Kind:     p.Kind,
+		APIGroup: p.APIGroup,
+		Name:     p.Name,
+	}
+
+	if p.Namespace != nil {
+		key.Namespace = *p.Namespace
+	}
+
+	if p.ParentPolicyID != nil {
+		key.ParentID = strconv.Itoa(*p.ParentPolicyID)
+	}
+
+	if p.SpecHash != nil {
+		key.SpecHash = *p.SpecHash
+	}
+
+	if p.Severity != nil {
+		key.Severity = *p.Severity
+	}
+
+	return key
```

Comment:
Similar comment to above, what do you think about generating a uniquely identifying string instead? I think this is less important in this case since you aren't allocating new slices for the conversion as you were for the parent policy.

Reply:
The pointers would still basically need to be handled this way, so I don't think changing it will really help.